### ISSUE TITLE
New Fortran Replacement Interface

### ISSUE THREAD

**Issue by naveen-rn** [openshmem-org/specification#194]
## Problem
Existing Fortran Interface in OpenSHMEM was deprecated in version-1.4. The reasons for the deprecation are available in #36. We are planning to completely remove the Fortran interface from the spec without any alternate replacement.   

## Proposed Solution
Replace the current include file based approach by adding a Fortran module. This SHMEM Fortran module would include a thin wrapper to directly call the C library routines. We could cast the incoming arguments if the data types are not supported in Fortran. A prototype version of the proposed SHMEM module is available in this [repository](https://github.com/naveen-rn/openshmem_fortran) with few example test cases to validate the wrapper.

## Effect
1. The new proposed SHMEM module-based replacement interface will have minimal effect on the current user code.
2. As the wrapper is entirely based on the C library routines, it would be easier to maintain from the standards perspective.  
3. From the implementer perspective, we could completely remove the Fortran callable C library interface from the implementation and just use the wrapper library.
4. We could also provide better integration with Coarray Fortran. Apart from just providing the regular wrappers, we could also define the interaction with allocatable coarray variables for allocating and deallocating symmetric arrays, update the library setup and exit operations.
5. This approach is slightly different from @jlinford [wrapper](https://github.com/jlinford/shmem_iso_c_binding) generated from TAU Performance systems. The definition of the wrapper routines in this proposal would enable us to directly replace the existing interface without much change in the user code.

## Example
In general the wrappers can be created as shown below:
``` Fortran
module shmem
    implicit none

    interface
        subroutine c_shmem_init() &
                   bind(c, name="shmem_init")
        end subroutine c_shmem_init
    end interface

    interface
        subroutine c_shmem_putmem(dest, src, nelems, pe) &
                   bind(c, name="shmem_putmem")
            use, intrinsic    :: iso_c_binding, only:c_int, c_size_t
            type(*),dimension(*)    :: dest, src
            integer(c_int),value    :: pe
            integer(c_size_t),value :: nelems
        end subroutine c_shmem_putmem
    end interface

contains:
    subroutine shmem_init()
        call c_shmem_init()
    end subroutine shmem_init

    subroutine shmem_putmem(dest, src, nelems, pe) 
        use, intrinsic :: iso_fortran_env, only:int32
        use, intrinsic :: iso_c_binding, only:c_int, c_size_t

        type(*),dimension(*),intent(in)   :: dest, src
        integer,intent(in)   :: nelems, pe
        integer(c_int)       :: c_pe
        integer(c_size_t)    :: c_nelems
        c_nelems = nelems
        c_pe     = pe

        call c_shmem_putmem(dest, src, c_nelems, c_pe)
    end subroutine shmem_putmem
end module shmem
```

**Comment by naveen-rn** [comment]
Since, #249, #250, #251, and #252 discusses the separate aspects of this proposal. We are closing any further comments on this issue and can be addressed on the separate issues. To close comments on this issue - we are attaching the slides used for the initial discussions. Also, only the comments are closed - the issue remains open until all the child issues are resolved.

[OpenSHMEM-Fortran.pptx](https://github.com/openshmem-org/specification/files/2435894/OpenSHMEM-Fortran.pptx)

**Comment by manjugv** [comment]
This issue is addressed via #251. Closing this one.
